feat(esp_http_client): Event to signal last header downloaded (IDFGH-15314)

bryghtlabs-richard opened the pull request:
## Description

Add an event to indicate when the HTTP response is ready to be received.

When parsing ND-JSON streams, this indicates the point when the ND-JSON stream is considered open, which occurs just after the last HTTP_EVENT_ON_HEADER.

ND-JSON stream clients cannot rely on the first HTTP_EVENT_ON_DATA, since that is only triggered by an event, which are optional and may never be sent, or only sent well after the last HTTP_EVENT_ON_HEADER.

Closes #15952

## Testing

Tested by connecting to an ND-JSON server, and receiving the new HTTP_EVENT_ON_HEADERS_DONE event before any ND-JSON line events were received.

If it's preferred to add to the end of `esp_http_client_event_id_t`, that's fine. I added it in roughly the order it would be expected to be received.

---

## Checklist

Before submitting a Pull Request, please ensure the following:

- [X] 🚨 This PR does not introduce breaking changes.
- [X] All CI checks (GH Actions) pass.
- [X] Documentation is updated as needed.
- [X] Tests are updated or added as necessary.
- [X] Code is well-commented, especially in complex areas.
- [X] Git history is clean — commits are squashed to the minimum necessary.

mahavirj commented on components/esp_http_client/include/esp_http_client.h:
```diff
@@ -39,6 +39,7 @@ typedef enum {
     HTTP_EVENT_HEADER_SENT = HTTP_EVENT_HEADERS_SENT, /*!< This header has been kept for backward compatibility
                                                            and will be deprecated in future versions esp-idf */
     HTTP_EVENT_ON_HEADER,       /*!< Occurs when receiving each header sent from the server */
+    HTTP_EVENT_ON_HEADERS_DONE, /*!< Occurs after all last header sent from server */
```
```suggestion
    HTTP_EVENT_ON_HEADERS_COMPLETE, /*!< Occurs when all headers are received on the client side */
```

mahavirj commented:
sha=e5e8bbcde4fa31d6148657563a662c1e7ea7e298

bryghtlabs-richard commented:
Thanks! I've updated it - I think I've addressed everything you asked, also updated the docs.

I built the examples & tests with my current IDF, a few didn't compile due to unrelated IDF version issues. Is there a place I can see the CI results you see?

nileshkale123 commented:
sha=30758d9beb7b317e49d4a8578c920100591fb973
